timerutil: remove Read field

RaduBerinde opened the pull request:
Note: first commit is #144335

The workaround to keep track of whether we read from the channel is no
longer necessary as of Go 1.23. Remove the `Timer.Read` field and
`MarkRead()` method.

Fixes #143655
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144207)
<!-- Reviewable:end -->

rafiss reviewed (approved):
thanks for doing this! i was just about to post a thread in slack about this since i had noticed https://github.com/golang/go/issues/37196 was closed.

i just had a small nit and question

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 4 of 94 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144207#-:-ONrOHR1A1s2Hbqsuk1I:b-6lr3do)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aerfrei, @angles-n-daemons, @DarrylWong, @DrewKimball, @herkolategan, @kev-cao, and @tbg)

___
*[`pkg/kv/kvserver/storeliveness/transport.go` line 369 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144207#-ONrMP3x3YcNjwvabQ1T:-ONrMP3x3YcNjwvabQ1U:b7d0n1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc44d41f3b66a46b384e7918a01208b53f3ea8ff/pkg/kv/kvserver/storeliveness/transport.go#L369)):*
> ```Go
> 					batchTimer.Read = true
> 				}
> 				break
> ```

I wonder if this `break` would be easier to understand if the `for` loop had a label on it, and the `break` were moved to be in the `case`, similar to this: https://github.com/cockroachdb/cockroach/blob/cffa347bea230cfb8dfcd41f13adabe22b6be8a0/pkg/kv/kvserver/client_raft_test.go#L1990-L1994

lgtm either way, but just something to consider in case you feel it helps improve legibility
___
*[`pkg/util/timeutil/timer.go` line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/144207#-ONrNIFnAadM5yM3_zxv:-ONrNIFnAadM5yM3_zxw:bla1k8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6ffcc8c2919b6e2f7b6cd711e88722bb5fb6fb3/pkg/util/timeutil/timer.go#L39)):*
> ```Go
> //	}
> //
> // Note that unlike the standard library's Timer type, this Timer will
> ```

i think we should still keep this paragraph beginning with "Note that unlike the standard library's Timer type..." since that is an important difference between this and `time.Timer`


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
timerutil: remove Read field

The workaround to keep track of whether we read from the channel is no
longer necessary as of Go 1.23. Remove the `Timer.Read` field and
`MarkRead()` method.

Fixes #143655
Release note: None

RaduBerinde committed:
kv: don't depend on timer.Read value

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144207)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aerfrei, @angles-n-daemons, @DarrylWong, @DrewKimball, @herkolategan, @kev-cao, and @rafiss)

___
*[`pkg/kv/kvserver/storeliveness/transport.go` line 369 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144207#-ONrMP3x3YcNjwvabQ1T:-ONu9UCLCIgEbUFGSCNV:b-tkfujg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc44d41f3b66a46b384e7918a01208b53f3ea8ff/pkg/kv/kvserver/storeliveness/transport.go#L369)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

I wonder if this `break` would be easier to understand if the `for` loop had a label on it, and the `break` were moved to be in the `case`, similar to this: https://github.com/cockroachdb/cockroach/blob/cffa347bea230cfb8dfcd41f13adabe22b6be8a0/pkg/kv/kvserver/client_raft_test.go#L1990-L1994

lgtm either way, but just something to consider in case you feel it helps improve legibility
</blockquote></details>

Not crazy about either honestly, I switched back to using a `bool` (but using a local variable).
___
*[`pkg/util/timeutil/timer.go` line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/144207#-ONrNIFnAadM5yM3_zxv:-ONuA5A-6tVRzJE9tGEI:b-4c5ezp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6ffcc8c2919b6e2f7b6cd711e88722bb5fb6fb3/pkg/util/timeutil/timer.go#L39)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think we should still keep this paragraph beginning with "Note that unlike the standard library's Timer type..." since that is an important difference between this and `time.Timer`
</blockquote></details>

Good catch, done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!





<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTRs!

bors r+

RaduBerinde commented:
bors r-

RaduBerinde commented:
bors r+
